src/platforms/wayland/displayclient.cpp

```diff
@@ -224,117 +283,76 @@ void mgw::DisplayClient::Output::mode(
-    dcout.current_format = mir_pixel_format_xrgb_8888;
-    dcout.connected = true;
-    dcout.used = true;
-    dcout.power_mode = mir_power_mode_on;
-    dcout.orientation = MirOrientation::mir_orientation_normal;
-    dcout.scale = 1.0;
-    dcout.form_factor = MirFormFactor::mir_form_factor_monitor;
-    dcout.gamma_supported = MirOutputGammaSupported::mir_output_gamma_unsupported;
-}
-
-mgw::DisplayClient::Output::~Output()
-{
-    if (output)
-    {
-        wl_output_destroy(output);
-    }
-
-    if (shell_toplevel)
-    {
-        xdg_toplevel_destroy(shell_toplevel);
-    }
-
-    if (shell_surface)
-    {
-        xdg_surface_destroy(shell_surface);
-    }
-
-    wl_surface_destroy(surface);
-
-    if (eglsurface != EGL_NO_SURFACE)
-    {
-        eglDestroySurface(owner->egldisplay, eglsurface);
-    }
-
-    if (eglctx != EGL_NO_CONTEXT)
-    {
-        eglDestroyContext(owner->egldisplay, eglctx);
-    }
-}
-
-
-// If building against newer Wayland protocol definitions we may miss trailing fields
-#pragma GCC diagnostic push
-#pragma GCC diagnostic ignored "-Wmissing-field-initializers"
-wl_output_listener const mgw::DisplayClient::Output::output_listener = {
-    &geometry,
-    &mode,
-    &done,
-    &scale,
-};
-#pragma GCC diagnostic pop
-
-void mgw::DisplayClient::Output::done(void* data, struct wl_output* /*wl_output*/)
-{
-    auto output = static_cast<Output*>(data);
-    output->on_done(*output);
-}
-
-void mgw::DisplayClient::Output::for_each_display_buffer(std::function<void(DisplayBuffer & )> const& f)
+void mgw::DisplayClient::Output::done()
 {
     if (!shell_surface)
     {
         static xdg_surface_listener const shell_surface_listener{
-            [](void*, auto xdg_surface, auto serial) { xdg_surface_ack_configure(xdg_surface, serial); },
+            [](void* self, auto, auto... args) { static_cast<Output*>(self)->surface_configure(args...); },
         };
         shell_surface = xdg_wm_base_get_xdg_surface(owner->shell, surface);
         xdg_surface_add_listener(shell_surface, &shell_surface_listener, this);
 
         static xdg_toplevel_listener const shell_toplevel_listener{
-            [](auto...){},
+            [](void* self, auto, auto... args) { static_cast<Output*>(self)->toplevel_configure(args...); },
             [](auto...){},
         };
         shell_toplevel = xdg_surface_get_toplevel(shell_surface);
         xdg_toplevel_add_listener(shell_toplevel, &shell_toplevel_listener, this);
 
         xdg_toplevel_set_fullscreen(shell_toplevel, output);
         wl_surface_set_buffer_scale(surface, round(dcout.scale));
-        wl_display_dispatch(owner->display);
+        wl_surface_commit(surface);
 
-        auto const& size = dcout.modes[dcout.current_mode_index].size;
+        // We call on_done() in the surface configure event handler
+        while (!surface_has_been_configured)
+        {
+            wl_display_roundtrip(owner->display);
+        }
```

Comment:
Do we need this both here and in the constructor? I would expect `surface_configure()` to be called before `done()`